akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -74,6 +74,10 @@ private[http] object StreamUtils {
         promise.failure(cause)
         ctx.fail(cause)
       }
+      override def onDownstreamFinish(ctx: Context[T]): TerminationDirective = {
```

Comment:
hm, why not just override postStop then and use `promise.trySuccess(())` (because `failure` might have been called already.).